018-cli-2.0/018-cli-2.0.md

```diff
@@ -0,0 +1,431 @@
+
+
+# Usage 3
+appwrite init function
+
+Give your function an ID
+function_three
+
+Give your function a name:
+FunctionThree
+
+Select a runtime:
+python-3.9
+
+Initializing function X ..
+```
+
+Under the hood, this command does the following
+
+1. Clones a publicly hosted git repository and copy's a particular folder into your working directory. This will allow us to fix / modify templates without creating a new CLI release 👍 
+
+2. Makes an entry in `appwrite.json`
+
+```json
+{
+    "project" : "xyz",
+    "functions": [
+        {
+          "name" : "My awesome function 💪",
+          "path" : "./FunctionOne",
+          "command" : "python main.py",
+          "runtime": "python3.9"
+        },
+    ],
+    "hosting": []
+}
+```
+## Project Structure
+---
+
+
+### Reader.php
+```php
+abstract class Reader {
+
+  protected string $path;
+  
+  protected array $data;
+
+  function read(): bool {}
+
+  function write(): bool {}
+   
+  function getPath():string {}
+
+  function setPath(string $path): void {}
+
+  function getProperty(string $key, mixed $default = null): mixed {}
+
+  function setProperty(string $key, mixed $value): void {}
+}
+```
+
+### Preference.php
+
+```php
+// This class models the hidden preferences.json file 
+class Preference extends Reader {
+
+  function getKey(): string {}
+  
+  function setKey(string $key): void {}
+
+  function getEndpoint(): string {}
+  
+  function setEndpoint(string $endpoint): void {}
+
+  function getCookie(): string {}
+  
+  function setCookie(string $cookie): void {}
+
+  function getLocale(): string {}
+
+  function setLocale(string $locale): void {}
+  
+  function isLoaded(): bool {}
+}
+
+```
+
+### Config.php
+
+
+```php
+// This class models the appwrite.json file.
+class Config extends Reader {
+
+  function getFunction(string $name): array {}
+
+  function getFunctions(): array {}
+  
+  function setFunction(array $function): void {}
+
+  function getProject(): string {}
+
+  function setProject(string $id): void {}
+
+}
+```
+
+
+
+### Prior art
+
+[prior-art]: #prior-art
+
+<!--
+
+Discuss prior art, both the good and the bad, in relation to this proposal. A
+few examples of what this can include are:
+
+- Does this functionality exist in other software and what experience has their
+  community had?
+- For other teams: What lessons can we learn from what other communities have
+  done here?
+- Papers: Are there any published papers or great posts that discuss this? If
+  you have some relevant papers to refer to, this can serve as a more detailed
+  theoretical background.
+
+This section is intended to encourage you as an author to think about the
+lessons from other software, provide readers of your RFC with a fuller picture.
+If there is no prior art, that is fine - your ideas are interesting to us
+whether they are brand new or if it is an adaptation from other software.
+
+Write your answer below.
+-->
+
+This RFC is inspired by the Firebase CLI.
+
```

Comment:
Does the Fission CLI have any insights to offer?
https://docs.fission.io/docs/usage/function/functions/